test/powershell/Language/Parser/Parser.Tests.ps1

```diff
@@ -648,39 +648,50 @@ foo``u{2195}abc
+            #Test various numbers using the standard notation.
             @{ Script = "0"; Expected = "0" }
-			@{ Script = "-2"; Expected = "-2" }
-			@{ Script = "2"; Expected = "2" }
-			@{ Script = $([int32]::MaxValue); Expected = $([int32]::MaxValue) }
-			@{ Script = $([int32]::MinValue); Expected = $([int32]::MinValue) }
-			#Tests for hexadecimal notation.
-			@{ Script = "0x0"; Expected = "0" }
-			@{ Script = "0xF"; Expected = "15" }
-			@{ Script = "0x80000000"; Expected = $([int32]::MinValue) }
-			@{ Script = "0xFFFFFFFF"; Expected = "-1" }
-			@{ Script = "0x7fffffff"; Expected = $([int32]::MaxValue) }
-			@{ Script = "0x100000000"; Expected = [int64]0x100000000 }
-			#Tests for exponential notation.
-			@{ Script = "0e0"; Expected = "0" }
-			@{ Script = "0e1"; Expected = "0" }
-			@{ Script = "1e2"; Expected = "100" }
-			@{ Script = $([int32]::MaxValue); Expected = $([int32]::MaxValue) }
-			@{ Script = "0e2"; Expected = "0" }
-			@{ Script = "-2e2"; Expected = "-200" }
-			@{ Script = "-0e2"; Expected = "0" }
-			@{ Script = "3e0"; Expected = "3" }
-			#Tests for floating point notation.
-			@{ Script = ".01"; Expected = "0.01" }
-			@{ Script = "0.0"; Expected = "0" }
-			@{ Script = "-0.1"; Expected = "-0.1" }
-			@{ Script = "9.12"; Expected = "9.12" }
-			@{ Script = $([single]::MinValue); Expected = $([float]::MinValue).ToString() }
-			@{ Script = $([float]::MaxValue); Expected = $([float]::MaxValue).ToString() }
-			#Tests for the K suffix for numbers.
-			@{ Script = "0kb"; Expected = "0" }
-			@{ Script = "1kb"; Expected = "1024" }
-			@{ Script = "-2KB"; Expected = "-2048" }
+            @{ Script = "-2"; Expected = "-2" }
+            @{ Script = "2"; Expected = "2" }
+            @{ Script = $([int32]::MaxValue); Expected = $([int32]::MaxValue) }
+            @{ Script = $([int32]::MinValue); Expected = $([int32]::MinValue) }
+            #Tests for hexadecimal notation.
+            @{ Script = "0x0"; Expected = "0" }
+            @{ Script = "0xF"; Expected = "15" }
+            @{ Script = "0x80000000"; Expected = $([int32]::MinValue) }
+            @{ Script = "0xFFFFFFFF"; Expected = "-1" }
+            @{ Script = "0x7fffffff"; Expected = $([int32]::MaxValue) }
+            @{ Script = "0x100000000"; Expected = [int64]0x100000000 }
+            #Tests for uint notation.
+            @{ Script = "0u"; Expected = [UInt32]::MinValue }
+            @{ Script = "15u"; Expected = "15" }
+            @{ Script = "1.5u"; Expected = "2" }
+            @{ Script = "32.1u"; Expected = "32" }
+            @{ Script = "4294967295u"; Expected = $([UInt32]::MaxValue) }
+            @{ Script = "401ul"; Expected = "401" }
+            # Banker's rounding on casting, of course.
+            @{ Script = "1024.5ul"; Expected = "1024" }
+            @{ Script = "4294967301u"; Expected = "4294967301" }
+            @{ Script = "18446744073709551615u"; Expected = $([UInt64]::MaxValue)}
```

Comment:
Can you add tests for the following cases please:
- `short` literals
- `short unsigned` literals
- `decimal` literals if the tests don't already exist (I know decimals were already supported, but if we're touching them we should test them)
- Various barewords that are "almost" number literals:
   + Numbers that ends with a non-numerical suffix like `7p`
   + Numbers that have a `u` but then have a non-numerical suffix like `20ux`
   + Numbers that use numerical literals but in a bad order or combination like `300ls` or `78su`
- Ideally we'd have some extra tests for case-insensitivity too, but they are less crucial.

I know it's onerous, but the tokenizer is the kind of place where good coverage is crucial. Having read through the code it all looks pristine, but testing like this is most important for picking up regressions later on

Reply:
I have designed and interactively tested that incorrectly ordered literals would fail. I'm not sure... how that particular test will respond; the tokenizer throws a 'not a valid numeric literal' when those are entered.

Should I create a new block of tests for ones that throw?

Reply:
Yeah that test will just flat fail for a failed conversion. Looks like there are two options:
- Turn the body into a `try`/`catch` so that it "returns" the error to compare to.
- Or, more cleanly, write a new test batch for tokenizer errors. The second is also nice because it will encourage that in future, and error cases are just as important to test in my view.

Reply:
Option 2 it is. :)

I'll probably leave off with just testing the tokens I've added and combinations thereof for now, but there'll be plenty of room for adding tests to ensure the others fail when they ought as well. :)

Reply:
@rjmholt turns out you have to try really hard to hide parse errors from the parser that you *want* to throw 😄 

Had to go from string -> script block -> invoke that *in a script block* -> pass to `| Should -Throw`. Otherwise it throws before it gets passed into `Should`!

Reply:
Yeah, it's a classic gotcha — you can't execute script before you parse it. Same reason why you can't use `Import-Module` to import PowerShell classes to use with other classes.

So you have to parse it at runtime with something like `[scriptblock]::Create("program")` or `[System.Management.Automation.Language.Parser]::Parse(...)`

Reply:
We have `[system.management.automation.psparser]::Tokenize`.